narwhals/narwhalify.py

```diff
@@ -0,0 +1,123 @@
+from __future__ import annotations
+
+from functools import partial
+from functools import wraps
+from typing import TYPE_CHECKING
+from typing import Any
+from typing import Callable
+from typing import TypeVar
+
+from narwhals.translate import from_native
+from narwhals.translate import to_native
+
+if TYPE_CHECKING:
+    from typing_extensions import Concatenate  # pragma: no cover
+    from typing_extensions import ParamSpec  # pragma: no cover
+
+    T = TypeVar("T")
+    PS = ParamSpec("PS")
+
+
+def narwhalify(
+    func: Callable[Concatenate[T, PS], T] | None = None,
+    from_kwargs: dict[str, Any] | None = None,
+    to_kwargs: dict[str, Any] | None = None,
+) -> Callable[Concatenate[Any, PS], Any] | Callable[[Any], Any]:
+    """Decorator that wraps a dataframe agnostic function between `from_native` and `to_native` conversion.
+
+    All the expressions used within `func` must be `narwhals` compatible, and we assume that only one
+    dataframe/series-like is returned from it.
+
+    Warning:
+        `func` can take an arbitrary number of dataframe/series objects which will be converted using `from_native`.
+        However, as these arguments are parsed as positional, make sure to:
+
+        - Pass them as the first positional arguments.
+        - Explicitly pass any other argument as keyword.
+
+        Check examples section below to see an instance of that.
```

Comment:
this seems a little dangerous, are sure people will use it correctly?

maybe the decorator could just deal with the simplest case of:
- single frame input
- *args
- **kwargs

?

Reply:
I don't know why but I didn't get notified on this comment.
Yes that was the original _fairly_ simple use case mentioned in the [issue itself](#133).

I am ok with having the simple case rollout first

Reply:
yup, sorry for having initially suggested to make it more complicated 🙈